dartagnan/src/main/java/com/dat3m/dartagnan/witness/WitnessGraph.java

```diff
@@ -2,156 +2,207 @@
+        StringBuilder str = new StringBuilder();
+        str.append("<graph edgedefault=\"directed\">\n");
+        for (String attr : attributes.keySet()) {
+            str.append("  <data key=\"").append(attr).append("\">").append(attributes.get(attr)).append("</data>\n");
+        }
+        for (Node n : nodes) {
+            str.append(n.toXML());
+        }
+        for (Edge e : edges) {
+            str.append(e.toXML());
+        }
+        str.append("</graph>");
+        return str.toString();
+    }
+
+    private List<MemoryCoreEvent> getEventsFromEdge(Program program, Edge edge) {
+        return edge.hasCline() ?
+            program.getThreadEvents(MemoryCoreEvent.class).stream()
+                .filter(e -> e.hasMetadata(SourceLocation.class))
+                .filter(e -> e.getMetadata(SourceLocation.class).lineNumber() == edge.getCline())
+                .toList() :
+            Collections.emptyList();
+    }
+
+    public BooleanFormula encode(EncodingContext context) {
+        Program program = context.getTask().getProgram();
+        BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        List<BooleanFormula> enc = new ArrayList<>();
+        List<MemoryCoreEvent> previous = new ArrayList<>();
+        for (Edge edge : edges.stream().filter(Edge::hasCline).toList()) {
+            List<MemoryCoreEvent> events = getEventsFromEdge(program, edge);
+            if (!previous.isEmpty() && !events.isEmpty()) {
+                enc.add(bmgr.or(Lists.cartesianProduct(previous, events).stream()
+                        .map(p -> context.edgeVariable("hb", p.get(0), p.get(1)))
+                        .toArray(BooleanFormula[]::new)));
+            }
+            previous = events;
+        }
+        return bmgr.and(enc);
+    }
+
+    private boolean graphEdgeImpliesHbEdge() {
+        return hasAttributed(PRODUCER.toString()) && getAttributed(PRODUCER.toString()).equals("Dartagnan");
+    }
+
+    public EventGraph getReadFromKnowledge(Program program, AliasAnalysis alias) {
+        EventGraph k = new EventGraph();
+        MemoryCoreEvent current = null;
+        MemoryCoreEvent last = null;
+        List<MemoryCoreEvent> currents;
+        for (Edge e : getPathToViolation()) {
+            currents = getEventsFromEdge(program, e);
+            current = currents.size() == 1 ? currents.get(0) : null;
+            // If a graph edge implies a hb-relation, inter-thread communication guarantees
+            // same address and thus rf.
+            if (last != null && current != null && last instanceof Store && current instanceof Load
+                    && ((graphEdgeImpliesHbEdge() && !last.getThread().equals(current.getThread()))
+                            || alias.mustAlias(last, current))) {
+                k.add(last, current);
+            }
+            last = current;
+        }
+        return k;
+    }
+
+    public EventGraph getCoherenceKnowledge(Program program, AliasAnalysis alias) {
```

Comment:
I think this method can be written a lot cleaner by first iterating over all edges to extract the stores into a list. This list will only contain stores and no `null`-entries.
Now you can do something like this:
```Java
for (int i = 0; i < writes.size() - 1; i++) {
   final Store w1 = writes.get(i);
   // Add co-edges to all later same-address writes.
   writes.sublist(i + 1, writes.size()).stream()
        .filter(w2 -> alias.mustAlias(w1, w2))
        .forEach(w2 -> must.add(w1, w2));
   // Special case for cross-thread edges.
   final Store immSucc = writes.get(i+1);
   if (graphEdgeImpliesHbEdge() && !w1.getThread().equals(w2.getThread()) {
      must.add(w1, immSucc);
  }
}
```